### PR TITLE
TH2-4842

### PR DESCRIPTION
**Pull request by lumber1000**
Secure connection support
Events batching
Extra metadata added to messages

### CODE REVIEW

File: src/main/kotlin/com/exactpro/th2/kafka/client/KafkaClientsFactory.kt [hunk 1]
```diff
@@ -31,27 +33,43 @@ import java.util.Properties
...
                 ConsumerConfig.RECONNECT_BACKOFF_MS_CONFIG to config.reconnectBackoffMs,
-                ConsumerConfig.RECONNECT_BACKOFF_MAX_MS_CONFIG to config.reconnectBackoffMaxMs
-            ))
+                ConsumerConfig.RECONNECT_BACKOFF_MAX_MS_CONFIG to config.reconnectBackoffMaxMs,
+
+                CommonClientConfigs.SECURITY_PROTOCOL_CONFIG to config.kafkaSecurityProtocol,
+                SaslConfigs.SASL_KERBEROS_SERVICE_NAME to config.kafkaSaslKerberosServiceName,
+                SaslConfigs.SASL_MECHANISM to config.kafkaSaslMechanism,
+                SaslConfigs.SASL_JAAS_CONFIG to config.kafkaSaslJaasConfig
+            )
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/kafka/client/KafkaClientsFactory.kt]: Maybe we should use `buildMap` instead and add SASL parameters only if they are set.

Also, I think we need validation for those properties (which of them is required for a secure connection?)

File: src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt [hunk 2]
```diff
@@ -181,6 +199,13 @@ class KafkaConnection(
         private val POLL_TIMEOUT = Duration.ofMillis(100L)
         private const val CONNECTIVITY_EVENT_TYPE = "ConnectivityServiceEvent"
 
+        private const val METADATA_TOPIC = "topic"
+        private const val METADATA_KEY = "key"
+        private const val METADATA_PARTITION = "partition"
+        private const val METADATA_OFFSET = "offset"
+        private const val METADATA_TIMESTAMP = "timestamp"
+        private const val METADATA_TIMESTAMP_TYPE = "timestampType"
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/kafka/client/KafkaConnection.kt]: I think we should make names more unique. Like `th2.kafka.topic`

File: src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt [hunk 3]
```diff
@@ -124,7 +125,20 @@ fun main(args: Array<String>) {
     LOGGER.info { "Microservice shutted down." }
 }
 
-class EventSender(private val eventRouter: MessageRouter<EventBatch>, private val rootEventId: EventID) {
+class EventSender(private val eventRouter: MessageRouter<EventBatch>, private val rootEventId: EventID) : AutoCloseable {
+
+    private val batchSenderExecutor = Executors.newSingleThreadScheduledExecutor()
+    private val eventBatcher = EventBatcher(
+        maxBatchSizeInBytes = 512 * 1024,
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt]: Should be taken from cradle configuration

Comment (lumber1000) [reply src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt]: getCradleConfiguration is depricated, getCradleNonConfidentialConfiguration() is 'protected'. Probably we should add these params to custom config.

**Review by OptimumCode — APPROVED** [review]

File: src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt [hunk 4]
```diff
@@ -61,6 +61,8 @@ fun main(args: Array<String>) {
         CommonFactory()
     }.apply { resources += "factory" to ::close }
 
+    LOGGER.info("java.security.krb5.conf = ${System.getProperty("java.security.krb5.conf")}")
```
Comment (OptimumCode) [inline src/main/kotlin/com/exactpro/th2/kafka/client/Main.kt]: I think that should be removed

**Pull request merged**